cpp/src/arrow/dataset/scanner.h

```diff
@@ -51,6 +52,8 @@ constexpr int64_t kDefaultBatchSize = 1 << 20;
 constexpr int32_t kDefaultBatchReadahead = 32;
 constexpr int32_t kDefaultFragmentReadahead = 8;
 
+using FragmentGenerator = std::function<Future<std::shared_ptr<Fragment>>()>;
```

Comment:
I think moving the subclass definitions means you can also move this alias and get rid of the async_generator.h include.

Reply:
I was able to move FragmentGenerator but `async_generator.h` was still needed for `Enumerated` which is a pity since `Enumerated` is small and self-contained.  Should I place it in its own file?

Reply:
Ah, that's alright then.